File: sortingAlgorithms/quickSort/quick_sort.py
def pivotHelper(arr, start, end):
    pivot = arr[start]
    idx = start

    for i in range(start, end):
        if arr[i] < pivot:
            idx += 1
            arr[idx], arr[i] = arr[i], arr[idx]

    arr[idx], arr[start] = arr[start], arr[idx]
    return idx


def quick_sort(arr, left, right):
    if left < right:
        idx = pivotHelper(arr, left, right)
        quick_sort(arr, left, idx)
        quick_sort(arr, idx + 1, right)

    return arr

arr = [1,5,7,4,1,6,8,7,9,2,3,7]
quick_sort(arr, 0, len(arr))
print(arr)

File: sortingAlgorithms/quickSort/quickSort.js
function pivotHelper(arr, start=0, end=arr.length) {
  const swap = (i, j) => ([arr[i], arr[j]] = [arr[j], arr[i]]);
  let pivot = arr[start];
  let swapIdx = start;

  for (let i = start + 1; i < end; i++) {
    if (pivot > arr[i]) {
      swapIdx++;
      swap(swapIdx, i);
    }
  }
  swap(start, swapIdx);
  return swapIdx;
}

function quickSort(arr, left=0, right=arr.length) {
  if (left < right) {
    let pivotIdx = pivotHelper(arr, left, right);
    quickSort(arr, left, pivotIdx);
    quickSort(arr, pivotIdx + 1, right);
  }
  return arr;
}

let arr = [5, 4, 9, 10, 2, 20, 8, 7, 3];
console.log(quickSort(arr));
console.log(arr);